File: backend/app/schemas/raw_material.py
from pydantic import BaseModel, ConfigDict, Field, model_validator

STATUS_PATTERN = "^(active|inactive|blocked)$"
MATERIAL_TYPE_PATTERN = "^(raw_material|packaging|consumable)$"


def _check_stock_thresholds(maximum_stock: float, reorder_point: float, safety_stock: float) -> None:
    # Only enforced once a maximum is actually set -- 0 (the default)
    # means "no ceiling configured yet", not "cap stock at zero".
    if maximum_stock <= 0:
        return
    if safety_stock > maximum_stock:
        raise ValueError("Safety stock cannot exceed maximum stock.")
    if reorder_point > maximum_stock:
        raise ValueError("Reorder point cannot exceed maximum stock.")


class RawMaterialCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    code: str = Field(min_length=1, max_length=30)
    name: str = Field(min_length=1, max_length=150)
    unit: str = Field(min_length=1, max_length=20)
    material_type: str = Field(default="raw_material", pattern=MATERIAL_TYPE_PATTERN)
    category: str | None = Field(default=None, max_length=100)
    description: str | None = None
    properties: dict[str, str] | None = None
    manufacturer: str | None = Field(default=None, max_length=150)
    manufacturer_part_number: str | None = Field(default=None, max_length=100)

    reorder_point: float = Field(default=0, ge=0)
    safety_stock: float = Field(default=0, ge=0)
    maximum_stock: float = Field(default=0, ge=0)
    storage_location: str | None = Field(default=None, max_length=100)

    default_supplier_id: int | None = Field(default=None, gt=0)
    unit_cost: float = Field(default=0, ge=0)

    inspection_required: bool = False
    certificate_required: bool = False
    qc_notes: str | None = None

    status: str = Field(default="active", pattern=STATUS_PATTERN)

    @model_validator(mode="after")
    def _check_stock_thresholds(self):
        _check_stock_thresholds(self.maximum_stock, self.reorder_point, self.safety_stock)
        return self


class RawMaterialUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    name: str | None = Field(default=None, min_length=1, max_length=150)
    unit: str | None = Field(default=None, min_length=1, max_length=20)
    material_type: str | None = Field(default=None, pattern=MATERIAL_TYPE_PATTERN)
    category: str | None = Field(default=None, max_length=100)
    description: str | None = None
    properties: dict[str, str] | None = None
    manufacturer: str | None = Field(default=None, max_length=150)
    manufacturer_part_number: str | None = Field(default=None, max_length=100)

    reorder_point: float | None = Field(default=None, ge=0)
    safety_stock: float | None = Field(default=None, ge=0)
    maximum_stock: float | None = Field(default=None, ge=0)
    storage_location: str | None = Field(default=None, max_length=100)

    default_supplier_id: int | None = Field(default=None, gt=0)
    unit_cost: float | None = Field(default=None, ge=0)

    inspection_required: bool | None = None
    certificate_required: bool | None = None
    qc_notes: str | None = None

    status: str | None = Field(default=None, pattern=STATUS_PATTERN)

    # A partial update only has the thresholds actually being changed --
    # full cross-field checking (against whatever isn't in this payload)
    # happens in RawMaterialCRUD.update, which has the existing row to
    # fill the gaps from before checking. This validator only catches the
    # case where the payload itself is internally inconsistent.
    @model_validator(mode="after")
    def _check_stock_thresholds(self):
        if self.maximum_stock is not None and self.maximum_stock > 0:
            if self.safety_stock is not None and self.safety_stock > self.maximum_stock:
                raise ValueError("Safety stock cannot exceed maximum stock.")
            if self.reorder_point is not None and self.reorder_point > self.maximum_stock:
                raise ValueError("Reorder point cannot exceed maximum stock.")
        return self


class RawMaterialOut(BaseModel):
    id: int
    code: str
    name: str
    unit: str
    material_type: str
    category: str | None
    description: str | None
    properties: dict[str, str] | None
    manufacturer: str | None
    manufacturer_part_number: str | None

    reorder_point: float
    safety_stock: float
    maximum_stock: float
    storage_location: str | None

    default_supplier_id: int | None
    unit_cost: float

    inspection_required: bool
    certificate_required: bool
    qc_notes: str | None

    status: str

    model_config = {"from_attributes": True}

File: frontend/src/lib/validation/rawMaterial.ts
import { z } from 'zod'
import { parseProperties } from './product'

// Raw materials use the identical free-form JSON spec approach as
// products (see backend/app/models/raw_material.py `properties`) --
// parseProperties is shared from lib/validation/product.ts rather than
// duplicated here. propertiesToInput is already re-exported from there
// via this directory's barrel (./index.ts); import it from '@/lib/
// validation/product' or '@/lib/validation' directly, not from here.

// Mirrors backend/app/schemas/raw_material.py.
export const rawMaterialSchema = z
  .object({
    code: z.string().trim().min(1, 'Code is required').max(30),
    name: z.string().trim().min(1, 'Name is required').max(150),
    unit: z.string().trim().min(1, 'Unit is required').max(20),
    material_type: z.enum(['raw_material', 'packaging', 'consumable']),
    category: z.string().trim().max(100).optional().or(z.literal('').transform(() => undefined)),
    description: z.string().optional().or(z.literal('').transform(() => undefined)),
    properties: z.string().optional().transform(parseProperties),
    manufacturer: z.string().trim().max(150).optional().or(z.literal('').transform(() => undefined)),
    manufacturer_part_number: z.string().trim().max(100).optional().or(z.literal('').transform(() => undefined)),

    reorder_point: z.coerce.number().min(0, 'Must be 0 or more'),
    safety_stock: z.coerce.number().min(0, 'Must be 0 or more'),
    maximum_stock: z.coerce.number().min(0, 'Must be 0 or more'),
    storage_location: z.string().trim().max(100).optional().or(z.literal('').transform(() => undefined)),

    default_supplier_id: z.coerce.number().int().positive().optional().or(z.literal('')),
    unit_cost: z.coerce.number().min(0, 'Must be 0 or more'),

    inspection_required: z.boolean(),
    certificate_required: z.boolean(),
    qc_notes: z.string().optional().or(z.literal('').transform(() => undefined)),

    status: z.enum(['active', 'inactive', 'blocked']),
  })
  .refine((v) => v.maximum_stock <= 0 || v.safety_stock <= v.maximum_stock, {
    message: 'Safety stock cannot exceed maximum stock.',
    path: ['safety_stock'],
  })
  .refine((v) => v.maximum_stock <= 0 || v.reorder_point <= v.maximum_stock, {
    message: 'Reorder point cannot exceed maximum stock.',
    path: ['reorder_point'],
  })

export type RawMaterialFormValues = z.input<typeof rawMaterialSchema>
export type RawMaterialSubmitValues = z.output<typeof rawMaterialSchema>

// RawMaterialOut round-trips every Update field, so the edit form only
// drops the immutable `code`.
export const rawMaterialEditSchema = rawMaterialSchema.omit({ code: true })

export type RawMaterialEditFormValues = z.input<typeof rawMaterialEditSchema>
export type RawMaterialEditSubmitValues = z.output<typeof rawMaterialEditSchema>
